server/src/main/java/keywhiz/service/daos/SecretDAO.java

```diff
@@ -385,6 +403,53 @@ public void deleteSecretsByName(String name) {
         .deleteSecretSeriesByName(name);
   }
 
+  /**
+   * Counts the number of secrets deleted before the specified cutoff.
+   *
+   * @param deleteBefore the cutoff date; secrets deleted before this date will be counted
+   */
+  public int countSecretsDeletedBeforeDate(DateTime deleteBefore) {
+    checkArgument(deleteBefore != null);
+
+    // identify the secrets deleted before this date
+    return secretSeriesDAOFactory.using(dslContext.configuration())
+        .getIdsForSecretSeriesDeletedBeforeDate(deleteBefore)
+        .size();
+  }
+
+  /**
+   * Permanently removes the series and all secrets-contents records ("versions") which were deleted
+   * before the given date.
+   *
+   * Unlike the "delete" endpoints above, THIS REMOVAL IS PERMANENT and cannot be undone by editing
+   * the database to restore the "current" entries.
+   *
+   * @param deleteBefore the cutoff date; secrets deleted before this date will be deleted
+   * @param sleepMillis how many milliseconds to sleep between each batch of removals
+   */
+  public void dangerPermanentlyRemoveSecretsDeletedBeforeDate(DateTime deleteBefore,
+      int sleepMillis) throws InterruptedException {
+    checkArgument(deleteBefore != null);
+    SecretSeriesDAO secretSeriesDAO = secretSeriesDAOFactory.using(dslContext.configuration());
+    SecretContentDAO secretContentDAO = secretContentDAOFactory.using(dslContext.configuration());
+
+    // identify the secrets deleted before this date
+    List<Long> ids = secretSeriesDAO.getIdsForSecretSeriesDeletedBeforeDate(deleteBefore);
+
+    // batch the list of secrets to be removed, to reduce load on the database
+    List<List<Long>> partitionedIds = Lists.partition(ids, MAX_ROWS_REMOVED_PER_TRANSACTION);
+    for (List<Long> idBatch : partitionedIds) {
+      // permanently remove the `secrets_contents` entries originally associated with these secrets
+      secretContentDAO.dangerPermanentlyRemoveRecordsForGivenSecretsIDs(idBatch);
```

Comment:
Should both the deletion operations occur within a transaction? If the first operation succeeds and the second fails, we're just left with secret series record(s) with no content, which I guess is not too bad? They would get deleted the next time the command is used.

Reply:
The next run would clean up the secret series, so I'm not too concerned.  Generally the two tables have been handled separately by their DAOs.